dev/set_matrix.py

```diff
@@ -0,0 +1,427 @@
+        help=(
+            "A string that represents a list of changed files in a pull request. "
+            "Valid only when `--diff-only` is specified."
+        ),
+    )
+
+    return parser.parse_args()
+
+
+def main():
+    args = parse_args()
+
+    print(divider("Parameters"))
+    print(json.dumps(vars(args), indent=2))
+
+    print(divider("Log"))
+    config = read_yaml(CONFIG_PATH)
+    try:
+        config_ref = read_yaml(args.ref_config)
+    except Exception as e:
+        print("Failed to read '{}' due to: '{}'".format(args.ref_config, e))
+        config_ref = {}
+
+    # assuming that the top-level keys in `ml-package-versions.yml` have the format:
+    # <flavor name>(-<suffix>) (e.g. sklearn, tensorflow-1.x, keras-tf1.x)
+    flavors = set(x.split("-")[0] for x in config.keys())
+    changed_flavors = get_changed_flavors(args.changed_files, flavors)
+
+    job_names = []
+    includes = []
+    should_include_all_items = not args.diff_only
+
+    for flavor, cfgs in config.items():
+        package_info = cfgs.pop("package_info")
+
+        should_include_all_items_in_this_flavor = (
+            should_include_all_items
+            or any(flavor.startswith(x) for x in changed_flavors)
+            or (flavor not in config_ref)
+            or (package_info != config_ref[flavor]["package_info"])
+        )
+
+        for key, cfg in cfgs.items():
+            if (
+                should_include_all_items_in_this_flavor
+                or (key not in config_ref[flavor])
+                or (cfg != config_ref[flavor][key])
+            ):
+                print("Processing {}.{}".format(flavor, key))
+
+                # released versions
+                versions = get_released_versions(package_info["pip_release"])
+                versions = filter_versions(
+                    versions, cfg["minimum"], cfg["maximum"], cfg.get("unsupported"),
+                )
+                versions = select_latest_micro_versions(versions)
+                for ver in versions:
+                    job_name = " / ".join([flavor, ver, key])
+                    job_names.append(job_name)
+                    requirements = process_requirements(cfg.get("requirements"), ver)
+                    install = "pip install -U '{}=={}'".format(package_info["pip_release"], ver)
+                    run = remove_comments(cfg["run"])
+                    includes.append(
+                        {
+                            "job_name": job_name,
+                            "requirements": requirements or None,
+                            "install": install,
+                            "run": run,
+                        }
+                    )
+
+                # development version
+                if "install_dev" in package_info:
+                    job_name = " / ".join([flavor, "dev", key])
+                    job_names.append(job_name)
+                    requirements = process_requirements(cfg.get("requirements"), "dev")
```

Comment:
Is it necessary to install dev versions of test requirements as well? The dev release of library X is unlikely to depend on behavior in the dev release of library Y, since the authors of X don't know when Y will be released. It's more likely that library X's behavior will interact with a currently-installable version of Y.

Let me know if I'm missing something, but I think we can drop this "dev" handling.

Reply:
Edit: Misinterpreted this. I think this is doing exactly what I described.